### PR TITLE
Extend meanshift tutorial

### PR DESCRIPTION
**Pull request by mehlukas**
Creates a general meanshift/camshift tutorial based on the [python version](https://docs.opencv.org/3.4.6/db/df8/tutorial_py_meanshift.html) of the tutorial. This is done according to the [Document improvement plan](https://github.com/opencv/opencv/wiki/Documentation-improvement-plan) and the discussion in #9520.
I tried to create a similar result as in the [background subtraction tutorial](https://github.com/opencv/opencv/blob/master/doc/tutorials/video/background_subtraction/background_subtraction.markdown) and as in #14314.

### This pullrequest changes

* extract python example code from [python tutorial on meanshift](https://github.com/opencv/opencv/blob/master/doc/py_tutorials/py_video/py_meanshift/py_meanshift.markdown) into the proper subfolder in samples
* add C++ example code
* add tutorial to the common tutorial folder with python and C++ code examples that can be toggled.

[Documentation Link](https://pullrequest.opencv.org/buildbot/export/pr/14393/docs/d7/d00/tutorial_meanshift.html)

```
force_builders_only=Docs
```

### CODE REVIEW

File: doc/tutorials/video/meanshift/meanshift.markdown [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+illustrated in the simple image given below:
+
+![image](images/meanshift_basics.jpg)
+
+The initial window is shown in blue circle with the name "C1". Its original center is marked in blue
+rectangle, named "C1_o". But if you find the centroid of the points inside that window, you will
+get the point "C1_r" (marked in small blue circle) which is the real centroid of window. Surely
+they don't match. So move your window such that circle of the new window matches with previous
+centroid. Again find the new centroid. Most probably, it won't match. So move it again, and continue
+the iterations such that center of window and its centroid falls on the same location (or with a
```
Comment (csukuangfj) [inline doc/tutorials/video/meanshift/meanshift.markdown]: (or with a ---> (or within a

**Comment by alalek** [comment]
Well done! Thank you for the contribution!

Lets avoid code/resources duplication:

1.
Please remove content of old the tutorial with adding a link on new one:

```
Tutorial content has been moved: @ref tutorial_meanshift
```

2.
Replace subpage in Python tutorials:

```.diff
--   @subpage tutorial_py_meanshift
+-   @subpage tutorial_meanshift
```

Take a look on commits here: https://github.com/alalek/opencv/commits/pr14393_r

**Comment by alalek** [comment]
Could you please describe what is broken by `@subpage`? (attach screenshot / provide document link)

**Comment by mehlukas** [comment]
It seems like the usage of `@subpage` leads to a wrong breadcrumb navigation in the header (see screenshot: Although the page is at the right spot, the navigation states `OpenCV-Python Tutorials > Video Analysis`). This probably happens, because the page was included twice with `@subpage`: From the python tutorials and from the central tutorials. Switching to `@ref` fixed it for me.
Screenshot_2019-05-01 OpenCV Meanshift and Camshift

**Comment by alalek** [comment]
@mehlukas Right! Thank you.

**Comment by alalek** [comment]
@catree Could you please take a quick look on this patch? (you have an expertise of writing these multi-language tutorials)

File: doc/tutorials/video/meanshift/meanshift.markdown [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+It applies meanshift first. Once meanshift converges, it updates the size of the window as,
+\f$s = 2 \times \sqrt{\frac{M_{00}}{256}}\f$. It also calculates the orientation of best fitting ellipse
+to it. Again it applies the meanshift with new scaled search window and previous window location.
+The process is continued until required accuracy is met.
+
+![image](images/camshift_face.gif)
+
+### Camshift in OpenCV
+
+It is almost same as meanshift, but it returns a rotated rectangle (that is our result) and box
```
Comment (catree) [inline doc/tutorials/video/meanshift/meanshift.markdown]:
```suggestion
It is similar to meanshift, but returns a rotated rectangle (that is our result) and box
```

File: doc/tutorials/video/meanshift/meanshift.markdown [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+Three frames of the result is shown below:
+
+![image](images/camshift_result.jpg)
+
+Additional Resources
+--------------------
+
+-#  French Wikipedia page on [Camshift](http://fr.wikipedia.org/wiki/Camshift). (The two animations
+    are taken from here)
```
Comment (catree) [inline doc/tutorials/video/meanshift/meanshift.markdown]:
```suggestion
    are taken from there)
```

File: doc/tutorials/video/meanshift/meanshift.markdown [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+2.  Bradski, G.R., "Real time face and object tracking as a component of a perceptual user
+    interface," Applications of Computer Vision, 1998. WACV '98. Proceedings., Fourth IEEE Workshop
+    on , vol., no., pp.214,219, 19-21 Oct 1998
+
+Exercises
+---------
+
+-#  OpenCV comes with a Python sample on interactive demo of camshift. Use it, hack it, understand
```
Comment (catree) [inline doc/tutorials/video/meanshift/meanshift.markdown]:
```suggestion
-#  OpenCV comes with a Python [sample](https://github.com/opencv/opencv/blob/3.4/samples/python/camshift.py) for an interactive demo of camshift. Use it, hack it, understand
```

File: doc/tutorials/video/table_of_content_video.markdown [hunk 5]
```diff
@@ -14,3 +14,9 @@ tracking and foreground extractions.
 
     We will learn how to extract foreground masks from both videos and sequences of images and
     to show them.
+
+-   @subpage tutorial_meanshift
+
+    *Languages:* Python, C++
+
+    Learn how to use the Meanshift and Camshift algorithms to find and track objects in videos.
```
Comment (catree) [inline doc/tutorials/video/table_of_content_video.markdown]:
```suggestion
    Learn how to use the Meanshift and Camshift algorithms to track objects in videos.
```

File: samples/cpp/tutorial_code/video/meanshift/camshift.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        rot_rect.points(points);
+        for (int i = 0; i < 4; i++)
+            line(frame, points[i], points[(i+1)%4], 255, 2);
+        imshow("img2", frame);
+
+        int keyboard = waitKey(30);
+        if (keyboard == 'q' || keyboard == 27)
+            break;
+    }
+    destroyAllWindows();
```
Comment (catree) [inline samples/cpp/tutorial_code/video/meanshift/camshift.cpp]:
```suggestion
```

Should not be needed I think.

File: samples/cpp/tutorial_code/video/meanshift/camshift.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    capture.release();
```
Comment (catree) [inline samples/cpp/tutorial_code/video/meanshift/camshift.cpp]:
```suggestion
```

Should not be needed also I think.

File: samples/cpp/tutorial_code/video/meanshift/meanshift.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+        // Draw it on image
+        rectangle(frame, track_window, 255, 2);
+        imshow("img2", frame);
+
+        int keyboard = waitKey(30);
+        if (keyboard == 'q' || keyboard == 27)
+            break;
+    }
+    destroyAllWindows();
```
Comment (catree) [inline samples/cpp/tutorial_code/video/meanshift/meanshift.cpp]:
```suggestion
```

File: samples/cpp/tutorial_code/video/meanshift/meanshift.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    capture.release();
```
Comment (catree) [inline samples/cpp/tutorial_code/video/meanshift/meanshift.cpp]:
```suggestion
```

File: samples/python/tutorial_code/video/meanshift/camshift.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        cv.imshow('img2',img2)
+
+        k = cv.waitKey(30) & 0xff
+        if k == 27:
+            break
+    else:
+        break
+
+cv.destroyAllWindows()
+cap.release()
```
Comment (catree) [inline samples/python/tutorial_code/video/meanshift/camshift.py]:
```suggestion
```

Should not be needed I think?

File: samples/python/tutorial_code/video/meanshift/camshift.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        img2 = cv.polylines(frame,[pts],True, 255,2)
+        cv.imshow('img2',img2)
+
+        k = cv.waitKey(30) & 0xff
+        if k == 27:
+            break
+    else:
+        break
+
+cv.destroyAllWindows()
```
Comment (catree) [inline samples/python/tutorial_code/video/meanshift/camshift.py]:
```suggestion
```

File: samples/python/tutorial_code/video/meanshift/meanshift.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        cv.imshow('img2',img2)
+
+        k = cv.waitKey(30) & 0xff
+        if k == 27:
+            break
+    else:
+        break
+
+cv.destroyAllWindows()
+cap.release()
```
Comment (catree) [inline samples/python/tutorial_code/video/meanshift/meanshift.py]:
```suggestion
```

Should not be needed I think?

File: samples/python/tutorial_code/video/meanshift/meanshift.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        img2 = cv.rectangle(frame, (x,y), (x+w,y+h), 255,2)
+        cv.imshow('img2',img2)
+
+        k = cv.waitKey(30) & 0xff
+        if k == 27:
+            break
+    else:
+        break
+
+cv.destroyAllWindows()
```
Comment (catree) [inline samples/python/tutorial_code/video/meanshift/meanshift.py]:
```suggestion
```

**Comment by alalek** [comment]
@catree Thank you!

`capture.release()` is not necessary in the program end.
`destroyAllWindows()` is not needed too. Both are harmless in general, but they are useless in modern OpenCV.

**Pull request merged**